test-network-function/accesscontrol/suite.go

```diff
@@ -130,16 +156,103 @@ func runTestOnPods(env *config.TestEnvironment, testCmd testcases.BaseTestCase,
+	if cmdOut == "" {
+		return map[string]string{}
+	}
+	lines := strings.Split(cmdOut, "\n")
+
+	crNamespaces := map[string]string{}
+	for _, line := range lines {
+		if line == "" {
+			// This is to filter the last line, which is always empty.
+			continue
+		}
+		lineFields := strings.Split(line, ",")
+		crNamespaces[lineFields[0]] = lineFields[1]
+	}
+
+	return crNamespaces
+}
+
+func testCrsNamespaces(crNames, configNamespaces []string) (invalidCrs map[string][]string) {
+	invalidCrs = map[string][]string{}
+	for _, crdName := range crNames {
+		getCrPluralNameCommand := fmt.Sprintf(ocGetCrPluralNameFormat, crdName)
+		crdPluralName := utils.ExecuteCommand(getCrPluralNameCommand, common.DefaultTimeout, common.GetContext(), func() {
+			tcClaimLogPrintf("CRD %s: Failed to get CR plural name.", crdName)
+		})
+		crNamespaces := getCrsNamespaces(crdName, crdPluralName)
+
+		ginkgo.By(fmt.Sprintf("CRD %s has %d CRs (plural name: %s).", crdName, len(crNamespaces), crdPluralName))
+		for crName, namespace := range crNamespaces {
+			ginkgo.By(fmt.Sprintf("Checking CR %s - Namespace %s", crName, namespace))
+			found := false
+			for _, configNamespace := range configNamespaces {
+				if namespace == configNamespace {
+					found = true
+				}
+			}
+
+			if !found {
+				tcClaimLogPrintf("CRD: %s (kind:%s) - CR %s has an invalid namespace (%s)", crdName, crdPluralName, crName, namespace)
+				if crNames, exists := invalidCrs[crdName]; exists {
+					invalidCrs[crdName] = append(crNames, crName)
+				} else {
+					invalidCrs[crdName] = []string{crName}
+				}
+			}
+		}
+	}
+	return invalidCrs
+}
+
 func testNamespace(env *config.TestEnvironment) {
-	ginkgo.When("test deployment namespace", func() {
+	ginkgo.When("test CNF namespaces", func() {
 		testID := identifiers.XformToGinkgoItIdentifier(identifiers.TestNamespaceBestPracticesIdentifier)
 		ginkgo.It(testID, func() {
-			for _, podUnderTest := range env.PodsUnderTest {
-				podName := podUnderTest.Name
-				podNamespace := podUnderTest.Namespace
-				ginkgo.By(fmt.Sprintf("Reading namespace of podnamespace= %s podname= %s, should not be 'default' or begin with openshift-", podNamespace, podName))
-				gomega.Expect(podNamespace).To(gomega.Not(gomega.Equal("default")))
-				gomega.Expect(podNamespace).To(gomega.Not(gomega.HavePrefix("openshift-")))
+			ginkgo.By(fmt.Sprintf("CNF resources' namespaces should not have any of the following prefixes: %v", invalidNamespacePrefixes))
+			var failedNamespaces []string
+			for _, namespace := range env.NameSpacesUnderTest {
+				ginkgo.By(fmt.Sprintf("Checking namespace %s", namespace))
+				for _, invalidPrefix := range invalidNamespacePrefixes {
+					if strings.HasPrefix(namespace, invalidPrefix) {
+						tcClaimLogPrintf("Namespace %s has invalid prefix %s", namespace, invalidPrefix)
+						failedNamespaces = append(failedNamespaces, namespace)
+					}
+				}
+			}
+			failedNamespacesNum := len(failedNamespaces)
+			if failedNamespacesNum > 0 {
```

Comment:
failedNamespacesNum is used once. it would be better to use len(failedNamespaces) directly